sys/net/application_layer/nanocoap/nanocoap.c

```diff
@@ -1059,6 +1069,54 @@ ssize_t coap_opt_add_proxy_uri(coap_pkt_t *pkt, const char *uri)
     return _add_opt_pkt(pkt, COAP_OPT_PROXY_URI, (uint8_t *)uri, strlen(uri));
 }
 
+ssize_t coap_opt_replace_etag(coap_pkt_t *pkt, const void *etag, size_t len)
+{
+    assert(len > 0 && len <= COAP_ETAG_LENGTH_MAX);
+    const coap_optpos_t *optpos = NULL;
+    uint8_t *opt = _get_option(pkt, COAP_OPT_ETAG, &optpos);
+    if (!opt) {
+        /* No etag, nothing modified */
+        return 0;
+    }
+
+    uint16_t delta = 0;
+    int opt_len = 0;
+    uint8_t *value = _parse_option(pkt, opt, &delta, &opt_len);
+
+    /* Sanity checking based on _parse_option return codes */
+    if (!value || opt_len < 0) {
+        /* Unable to parse option or packet, do nothing */
+        return 0;
+    }
+
+    /* Length supplied to the initial call must be at least as much as supplied to this call */
+    assert((size_t)opt_len >= len);
+
+    /* Between -7 and 0, inclusive */
+    ssize_t len_diff = len - opt_len;
+
+    if (len_diff != 0) {
+        /* shift everything */
+        /* addresses of the rest of the packet after the option */
+        uint8_t *rest_of_the_packet = value + opt_len;
+        uint8_t *end_of_the_packet = pkt->payload + pkt->payload_len;
+
+        uint16_t lastonum = COAP_OPT_ETAG - delta;
+        size_t new_len = coap_put_option(opt, lastonum, COAP_OPT_ETAG, etag, len);
+
+        /* Move the rest of the packet to right after the newly written option */
+        memmove(opt + new_len, rest_of_the_packet, end_of_the_packet - rest_of_the_packet);
+
+        /* Update the payload data */
+        pkt->payload += len_diff;
```

Comment:
The gut feeling says no, all pointer values here are grabbed from parse functions already in nanocoap.

The only way I see this happening is when the payload ptr is already incorrect or when the dummy etag value was shorter than the new one.

Reply:
> Sounds like there should be a check at least a `(pkt->payload + len_diff) < opt` check, just to avoid stupidity ;-)

API misuse should already be covered by the [assert(opt_len >= len)](https://github.com/RIOT-OS/RIOT/pull/19980/files#diff-1c737011eccee82809946c008c7b47851a01a64d6610761b539d1fc9098900a1R1093). I can add another assert, however I think this mostly results in sanity testing nanocoap internal functions. 

But maybe a second look from somebody else well versed in pointer magic might help here as maybe I'm missing an edge case.

Reply:
Yes, I think an assert would be enough in this case, just to make sure the pre-conditions are met.

Reply:
Added